### PR TITLE
Optional and Object

### PR DESCRIPTION
**Pull request by olynch**

### CODE REVIEW

**Review by fivegrant — COMMENTED** [review]
@olynch I'm still double checking a few things but I was wondering if you've noticed the JSON Schema validating on any JSON object. Would it break InterTypes to have the user specify which types get definitions? I think that might help a bit as to not stuff a bunch of definitions at the top level

File: src/intertypes/InterTypes.jl [hunk 1]
```diff
@@ -113,6 +116,36 @@ function generate_module(mod::Module, target::Type{<:ExportTarget}, path="."; ta
...
 end
 
+module InterTypeSupport
+using OrderedCollections
+using StructEquality
+import StructTypes
+export Object, Optional
+
+@struct_hash_equal struct Object{T}
+  fields::OrderedDict{Symbol, T}
```
Comment (jpfairbanks) [inline src/intertypes/InterTypes.jl:126]: Is this supposed to require that all the fields have same type?

File: src/intertypes/json.jl [hunk 2]
```diff
@@ -242,68 +263,75 @@ function write(io::IO, format::JSONFormat, acs::ACSet)
...
-      "type" => "string",
-      "contentEncoding" => "base64",
-      "\$comment" => "Binary"
+      :type => "string",
+      :contentEncoding => "base64",
+      Symbol("\$comment") => "Binary"
+    )
+    OptionalType(elemtype) => begin
+      schema = tojsonschema(elemtype)
+      schema[:type] = [schema[:type], "null"]
```
Comment (jpfairbanks) [inline src/intertypes/json.jl:317]: The schema of a type can be a list including the string null? Is this a JSONSchema idiom?

File: src/intertypes/InterTypes.jl (hunk 1 above)
Comment (olynch) [reply src/intertypes/InterTypes.jl:126]: Yes. We don't support dynamic types. If you want fields of different types, use a struct where each field can be configured individually.

File: src/intertypes/json.jl (hunk 2 above)
Comment (olynch) [reply src/intertypes/json.jl:317]: I think so? This is what showed up on the jsonschema website when I poked around for it, but I'm no jsonschema expert.

**Comment by olynch** [comment]
> @olynch I'm still double checking a few things but I was wondering if you've noticed the JSON Schema validating on any JSON object. Would it break InterTypes to have the user specify which types get definitions? I think that might help a bit as to not stuff a bunch of definitions at the top level

I'm not really sure what you mean by this with regard to "which types get definitions"?

File: src/intertypes/json.jl (hunk 2 above)
Comment (jpfairbanks) [reply src/intertypes/json.jl:317]: @fivegrant can probably confirm that it is what he would expect for a nullable field.

File: src/intertypes/InterTypes.jl (hunk 1 above)
Comment (jpfairbanks) [reply src/intertypes/InterTypes.jl:126]: ok, so you can have dynamic names where all the fields are the same type. Or static names with configurable types (also statically)

**Review by jpfairbanks — APPROVED** [review]

**Comment by fivegrant** [comment]
@olynch If you look at the [Petrinet Schema](https://github.com/DARPA-ASKEM/Model-Representations/blob/main/petrinet/petrinet_schema.json), you'll notice it only validates on object. InterTypes, however, produces a bunch of `$defs`: 
```json
"$defs": {
    "Parameter": {
      "type": "object",
      "properties": {
        "type": {
          "type": "string",
          "$comment": "Str"
        },
        "value": {
          "type": "number",
          "$comment": "F64"
        }
      },
      "required": [
        "type",
        "value"
      ]
    },
    "Condition": {
      "type": "object",
      "properties": {
        "type": {
          "type": "string",
          "$comment": "Str"
        },
        "value": {
          "type": "number",
          "$comment": "F64"
        },
        "domain_mesh": {
          "type": "string",
    ....
    "Thing": {
      "type": "object",
      "properties": {
        "a": {"type": "Condition"}, "b":{"type":"Parameter"}}
      }
    }
```

When validating against this schema with a bunch of `$defs`, many tools don't let you specify a particular one and see if a given JSON matches just one of the entries. For example, I just have to match one of the above to pass

**Comment by olynch** [comment]
> @olynch If you look at the [Petrinet Schema](https://github.com/DARPA-ASKEM/Model-Representations/blob/main/petrinet/petrinet_schema.json), you'll notice it only validates on object. InterTypes, however, produces a bunch of `$defs`:
> 
> ```json
> "$defs": {
>     "Parameter": {
>       "type": "object",
>       "properties": {
>         "type": {
>           "type": "string",
>           "$comment": "Str"
>         },
>         "value": {
>           "type": "number",
>           "$comment": "F64"
>         }
>       },
>       "required": [
>         "type",
>         "value"
>       ]
>     },
>     "Condition": {
>       "type": "object",
>       "properties": {
>         "type": {
>           "type": "string",
>           "$comment": "Str"
>         },
>         "value": {
>           "type": "number",
>           "$comment": "F64"
>         },
>         "domain_mesh": {
>           "type": "string",
>     ....
>     "Thing": {
>       "type": "object",
>       "properties": {
>         "a": {"type": "Condition"}, "b":{"type":"Parameter"}}
>       }
>     }
> ```
> 
> When validating against this schema with a bunch of `$defs`, many tools don't let you specify a particular one and see if a given JSON matches just one of the entries. For example, I just have to match one of the above to pass

This is a valid point, but not really relevant to this PR. Can you open an issue for it?

However, I would argue that this is a better way of doing it, because it maps closer to how one declares types in the programming languages that use intertypes.

**Review by fivegrant — APPROVED** [review]
@olynch I brought the `$defs` issue because I think adding objects made the JSON Schema noticeably more permissive but it's been a minute since I've validated previously

Anyway, I waited to review this PR until I made sure everything worked in the MR one *and it does* so LGTM!

**Comment by fivegrant** [comment]
small note: I added InterTypes so they match the examples exactly. Changes that require modifying the examples should be done in a different PR.

**Pull request merged**